pkg/config/setup/config.go

```diff
@@ -1765,38 +1768,108 @@ func ResolveSecrets(config pkgconfigmodel.Config, secretResolver secrets.Compone
 		}
 
-		err = secretResolver.ResolveWithCallback(
-			yamlConf,
-			origin,
-			func(yamlPath []string, value any) bool {
-				settingName := strings.Join(yamlPath, ".")
-
-				// We received a notification about an unknown setting. This means that the a value
-				// inside a map was updated (ie: settings like additional_endpoints,
-				// kubernetes_node_annotations_as_tags, ...). This is an issue when a setting use a
-				// map[string] as a value. The key can contain a '.' making it impossible to set or get.
-				//
-				// The secrets resolver doesn't have the notion of what is a known setting and what
-				// isn't. Returning false tells the secretResolver that we refuse the notification and
-				// we want to be notified for the parent key.
-				//
-				// See secrets.ResolveCallback documentation for more information.
-				//
-				if !config.IsKnown(settingName) {
-					return false
-				}
-
-				log.Debugf("replacing handle for setting '%s' with secret value", settingName)
-				config.Set(settingName, value, pkgconfigmodel.SourceAgentRuntime)
-				return true
-			})
-		if err != nil {
+		secretResolver.SubscribeToChanges(func(handle string, settingPath []string, oldValue, newValue any) {
+			if err := configAssignAtPath(config, settingPath, newValue); err != nil {
+				log.Errorf("could not assign to config: %s", err)
+			}
+		})
+		if _, err = secretResolver.Resolve(yamlConf, origin); err != nil {
 			return fmt.Errorf("unable to decrypt secret from datadog.yaml: %v", err)
 		}
 	}
 	return nil
 }
 
+// assign a value to the given setting of the config
+// This works around viper issues that prevent us from assigning to fields that have a dot in the
+// name (example: 'additional_endpoints.http://url.com') and also allows us to assign to individual
+// elements of a slice of items (example: 'proxy.no_proxy.0' to assign index 0 of 'no_proxy')
+func configAssignAtPath(config pkgconfigmodel.Config, settingPath []string, newValue any) error {
+	settingName := strings.Join(settingPath, ".")
+	if config.IsKnown(settingName) {
+		config.Set(settingName, newValue, pkgconfigmodel.SourceAgentRuntime)
+		return nil
+	}
+
+	// Trying to assign to an unknown config field can happen when trying to set a
+	// value inside of a compound object (a slice or a map) which allows arbitrary key
+	// values. Some settings where this happens include `additional_endpoints`, or
+	// `kubernetes_node_annotations_as_tags`, etc. Since these arbitrary keys can
+	// contain a '.' character, we are unable to use the standard `config.Set` method.
+	// Instead, we remove trailing elements from the end of the path until we find a known
+	// config field, retrieve the compound object at that point, and then use the trailing
+	// elements to figure out how to modify that particular object, before setting it back
+	// on the config.
+	//
+	// Example: configAssignAtPath(config, ['process_config', 'additional_endpoints', 'http://url.com', '0'], 'password')
+	//
```

Comment:
```suggestion
	// Example with the following configuration:
	//
	//    process_config:
	//      additional_endpoints:
	//        http://url.com:
	//         - ENC[handle to the password]
	//
	// this function willl be called with:
	//    configAssignAtPath(config, ['process_config', 'additional_endpoints', 'http://url.com', '0'], 'password')
	//
```
WDYT ?

Reply:
Definitely helps explain it better, thanks!